zipline/pipeline/engine.py

```diff
@@ -315,10 +329,13 @@ def compute_chunk(self, graph, dates, assets, initial_workspace):
             mask, mask_dates = self._mask_and_dates_for_term(
                 term, workspace, graph, dates
             )
+
             if term.atomic:
-                # FUTURE OPTIMIZATION: Scan the resolution order for terms in
-                # the same dataset and load them here as well.
-                to_load = [term]
+                to_load = sorted(
+                    atomic_groups[atomic_group_key(term)],
+                    key=lambda t: t.dataset
```

Comment:
Since the hierarchy is generally loader:dataset:column, I thought loaders would often want to group by dataset - I know this is the case for fundamentals.  Instead of re-sorting to enable a lazy groupby inside each loader, my though was to do it once.